src/Workspaces/CSharp/Portable/Indentation/CSharpIndentationService.Indenter.cs

```diff
@@ -4,40 +4,47 @@
 using System.Diagnostics;
 using System.Linq;
 using System.Threading;
-using Microsoft.CodeAnalysis.CSharp;
 using Microsoft.CodeAnalysis.CSharp.Extensions;
 using Microsoft.CodeAnalysis.CSharp.Formatting;
 using Microsoft.CodeAnalysis.CSharp.Syntax;
-using Microsoft.CodeAnalysis.Editor.Implementation.Formatting.Indentation;
 using Microsoft.CodeAnalysis.Formatting;
 using Microsoft.CodeAnalysis.Formatting.Rules;
+using Microsoft.CodeAnalysis.Indentation;
 using Microsoft.CodeAnalysis.LanguageServices;
 using Microsoft.CodeAnalysis.Options;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Text;
-using Microsoft.CodeAnalysis.Text.Shared.Extensions;
 using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.Editor.CSharp.Formatting.Indentation
+namespace Microsoft.CodeAnalysis.CSharp.Indentation
 {
     internal partial class CSharpIndentationService
     {
         internal class Indenter : AbstractIndenter
         {
             public Indenter(
-                ISyntaxFactsService syntaxFacts,
+                Document document,
                 SyntaxTree syntaxTree,
                 IEnumerable<AbstractFormattingRule> rules,
                 OptionSet optionSet,
                 TextLine line,
                 CancellationToken cancellationToken) :
-                base(syntaxFacts, syntaxTree, rules, optionSet, line, cancellationToken)
+                base(document, syntaxTree, rules, optionSet, line, cancellationToken)
             {
             }
 
-            public override bool ShouldUseFormatterIfAvailable()
+            protected override bool ShouldUseTokenIndenter(out SyntaxToken syntaxToken)
                 => ShouldUseSmartTokenFormatterInsteadOfIndenter(
-                    Rules, Root, LineToBeIndented, OptionSet, CancellationToken);
+                    Rules, Root, LineToBeIndented, OptionSet, out syntaxToken);
+
+            protected override ISmartTokenFormatter CreateSmartTokenFormatter()
+            {
+                var workspace = Document.Project.Solution.Workspace;
+                var formattingRuleFactory = workspace.Services.GetService<IHostDependentFormattingRuleFactoryService>();
+                var rules = formattingRuleFactory.CreateRule(Document, LineToBeIndented.Start).Concat(Formatter.GetDefaultFormattingRules(Document));
+
+                return new CSharpSmartTokenFormatter(OptionSet, rules, Root);
```

Comment:
token formatter actually change buffer and indenter is not supposed to touch buffer. I am not sure.. how two can be merged..

Reply:
That doesn't seem to be the case.  ISmartTokenFormatter has a single method, which is: ```Task<IList<TextChange>> FormatTokenAsync(...)```.  All it does is say what expected changes should be returned.  It's up to the feature calling this to decide what to do with this.